File: mev-programs/tests/merkle-tree.ts
import { createHash } from "crypto";

export const convertBufProofToNumber = (buffers: Buffer[]): number[][] => {
  return buffers.reduce((prev, cur) => {
    const numArray = new Array(cur.length);
    for (let i = 0; i < cur.length; i = i + 1) numArray[i] = cur[i];
    prev.push(numArray);
    return prev;
  }, [] as number[][]);
};

export class MerkleTree {
  leafs: Array<Buffer>;
  layers: Array<Array<Buffer>>;

  constructor(leafs: Array<Buffer>) {
    this.leafs = leafs.slice();
    this.layers = [];

    let hashes = this.leafs.map(MerkleTree.nodeHash);
    while (hashes.length > 0) {
      // console.log("Hashes", this.layers.length, hashes);
      this.layers.push(hashes.slice());
      if (hashes.length === 1) break;
      hashes = hashes.reduce((acc, cur, idx, arr) => {
        if (idx % 2 === 0) {
          const nxt = arr[idx + 1];
          acc.push(MerkleTree.internalHash(cur, nxt));
        }
        return acc;
      }, Array<Buffer>());
    }
  }

  static sha256(...args: Buffer[]): Buffer {
    const hash = createHash("sha256");
    args.forEach((x) => hash.update(x));
    return hash.digest();
  }

  static nodeHash(data: Buffer): Buffer {
    //normal method
    //return MerkleTree.sha256(Buffer.from([0x00]), data);

    //jito method since stuff comes in already hashed, see https://github.com/jito-labs/jito-solana/blob/efa56db23943ae1b2a940f1e215aa75edc99be16/merkle-tree/src/merkle_tree.rs#L57
    return MerkleTree.sha256(Buffer.from([0x00]), MerkleTree.sha256(data));
  }

  static internalHash(first: Buffer, second: Buffer | undefined): Buffer {
    if (!second) return first;
    const [fst, snd] = [first, second].sort(Buffer.compare);

    return MerkleTree.sha256(Buffer.from([0x01]), fst, snd);
  }

  getRoot(): Buffer {
    return this.layers[this.layers.length - 1][0];
  }

  getProof(idx: number): Buffer[] {
    return this.layers.reduce((proof, layer) => {
      const sibling = idx ^ 1;
      if (sibling < layer.length) {
        proof.push(layer[sibling]);
      }

      idx = Math.floor(idx / 2);

      return proof;
    }, []);
  }

  getHexRoot(): string {
    return this.getRoot().toString("hex");
  }

  getHexProof(idx: number): string[] {
    return this.getProof(idx).map((el) => el.toString("hex"));
  }

  verifyProof(idx: number, proof: Buffer[], root: Buffer): boolean {
    let pair = MerkleTree.nodeHash(this.leafs[idx]);
    for (const item of proof) {
      pair = MerkleTree.internalHash(pair, item);
    }

    return pair.equals(root);
  }

  static verifyClaim(leaf: Buffer, proof: Buffer[], root: Buffer): boolean {
    let pair = MerkleTree.nodeHash(leaf);
    for (const item of proof) {
      pair = MerkleTree.internalHash(pair, item);
    }

    return pair.equals(root);
  }
}

File: mev-programs/programs/priority-fee-distribution/tests/helpers/merkle_tree.rs
use solana_program::hash::hashv;

#[derive(Debug, Clone)]
pub struct MerkleTree {
    pub leafs: Vec<Vec<u8>>,
    pub layers: Vec<Vec<[u8; 32]>>,
}

impl MerkleTree {
    pub fn new(leafs: Vec<Vec<u8>>) -> Self {
        let mut layers: Vec<Vec<[u8; 32]>> = Vec::new();

        let mut hashes: Vec<[u8; 32]> = leafs.iter().map(|leaf| Self::node_hash(leaf)).collect();

        while !hashes.is_empty() {
            layers.push(hashes.clone());

            if hashes.len() == 1 {
                break;
            }

            let mut next_layer: Vec<[u8; 32]> = Vec::new();

            for i in (0..hashes.len()).step_by(2) {
                let first = hashes[i];
                let second = hashes.get(i + 1).copied();
                let combined = Self::internal_hash(first, second);
                next_layer.push(combined);
            }

            hashes = next_layer;
        }

        Self { leafs, layers }
    }

    #[inline]
    pub fn sha256(parts: &[&[u8]]) -> [u8; 32] {
        hashv(parts).to_bytes()
    }


    pub fn node_hash(data: &[u8]) -> [u8; 32] {
        //jito method since stuff comes in already hashed, see https://github.com/jito-foundation/jito-solana/blob/master/merkle-tree/src/merkle_tree.rs#L6
        let inner = Self::sha256(&[data]);
        Self::sha256(&[&[0x00], &inner]) 
    }

    pub fn internal_hash(first: [u8; 32], second: Option<[u8; 32]>) -> [u8; 32] {
        let Some(second) = second else {
            return first;
        };

        let (fst, snd) = if first <= second {
            (first, second)
        } else {
            (second, first)
        };

        Self::sha256(&[&[0x01], &fst, &snd])
    }

    pub fn get_root(&self) -> [u8; 32] {
        self.layers.last().expect("Merkle tree has no layers")[0]
    }

    pub fn get_proof(&self, mut idx: usize) -> Vec<[u8; 32]> {
        let mut proof = Vec::new();

        for layer in &self.layers {
            let sibling = idx ^ 1;

            if sibling < layer.len() {
                proof.push(layer[sibling]);
            }

            idx /= 2;
        }

        proof
    }

    pub fn verify_proof(&self, idx: usize, proof: &[[u8; 32]], root: [u8; 32]) -> bool {
        let mut pair = Self::node_hash(&self.leafs[idx]);

        for p in proof {
            pair = Self::internal_hash(pair, Some(*p));
        }

        pair == root
    }

    pub fn verify_claim(leaf: &[u8], proof: &[[u8; 32]], root: [u8; 32]) -> bool {
        let mut pair = Self::node_hash(leaf);

        for p in proof {
            pair = Self::internal_hash(pair, Some(*p));
        }

        pair == root
    }
}
